broker/src/main/java/io/camunda/zeebe/broker/transport/ApiRequestHandler.java

```diff
@@ -0,0 +1,147 @@
+ */
+package io.camunda.zeebe.broker.transport;
+
+import io.camunda.zeebe.broker.Loggers;
+import io.camunda.zeebe.broker.transport.ApiRequestHandler.RequestReader;
+import io.camunda.zeebe.broker.transport.ApiRequestHandler.ResponseWriter;
+import io.camunda.zeebe.transport.RequestHandler;
+import io.camunda.zeebe.transport.ServerOutput;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferReader;
+import io.camunda.zeebe.util.buffer.BufferWriter;
+import io.camunda.zeebe.util.sched.Actor;
+import org.agrona.DirectBuffer;
+import org.agrona.sbe.MessageDecoderFlyweight;
+import org.slf4j.Logger;
+
+/**
+ * A {@link RequestHandler} that automatically decodes requests and encodes successful and error
+ * responses.
+ *
+ * @param <R> a {@link RequestReader} that reads the request
+ * @param <W> a {@link ResponseWriter} that writes the response
+ */
+public abstract class ApiRequestHandler<R extends RequestReader<?>, W extends ResponseWriter>
+    extends Actor implements RequestHandler {
+
+  public static final Logger LOG = Loggers.TRANSPORT_LOGGER;
+  private final ErrorResponseWriter errorResponseWriter = new ErrorResponseWriter();
+  private final R requestReader;
+  private final W responseWriter;
+
+  protected ApiRequestHandler(final R requestReader, final W responseWriter) {
+    this.requestReader = requestReader;
+    this.responseWriter = responseWriter;
+  }
+
+  /**
+   * Handles a request. The {@param requestReader} is populated at this point.
+   *
+   * @param partitionId the current partition id
+   * @param requestId the current request id
+   * @param requestReader a {@link R reader} that is already populated with the request data
+   * @param responseWriter a {@link W writer} that can be used to write successful responses
+   * @param errorWriter a {@link ErrorResponseWriter} that can be used to write error responses
+   * @return a {@link ErrorResponseWriter} when the handling failed, or a {@link W ResponseWriter}
+   *     when the handling was successful. Writes to the other writer will be ignored and are not
+   *     sent as a response.
+   */
+  protected abstract Either<ErrorResponseWriter, W> handle(
+      int partitionId,
+      long requestId,
+      R requestReader,
+      W responseWriter,
+      ErrorResponseWriter errorWriter);
+
+  @Override
+  public final void onRequest(
+      final ServerOutput serverOutput,
+      final int partitionId,
+      final long requestId,
+      final DirectBuffer buffer,
+      final int offset,
+      final int length) {
+    actor.submit(() -> handleRequest(serverOutput, partitionId, requestId, buffer, offset, length));
+  }
+
+  private void handleRequest(
+      final ServerOutput serverOutput,
+      final int partitionId,
+      final long requestId,
+      final DirectBuffer buffer,
+      final int offset,
+      final int length) {
+    try {
+      requestReader.wrap(buffer, offset, length);
```

Comment:
:x:  Here, I think, there won't be any exception if CommandApi receives unknown requests types.

This makes me think that we might need to add tests for CommandApiRequestHandler/CommandApiRequestReader.

Reply:
There is an exception:

```
15:24:12.693 [io.camunda.zeebe.broker.transport.commandapi.CommandApiRequestHandler] [-zb-actors-0] ERROR io.camunda.zeebe.broker.transport - Failed to deserialize message
java.lang.IllegalStateException: Invalid TEMPLATE_ID: 65535
	at io.camunda.zeebe.protocol.record.ExecuteCommandRequestDecoder.wrapAndApplyHeader(ExecuteCommandRequestDecoder.java:94) ~[classes/:?]
	at io.camunda.zeebe.broker.transport.commandapi.CommandApiRequestReader.wrap(CommandApiRequestReader.java:63) ~[classes/:?]
	at io.camunda.zeebe.broker.transport.ApiRequestHandler.handleRequest(ApiRequestHandler.java:81) ~[classes/:?]
	at io.camunda.zeebe.broker.transport.ApiRequestHandler.lambda$onRequest$0(ApiRequestHandler.java:70) ~[classes/:?]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:73) [classes/:?]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [classes/:?]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) [classes/:?]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:95) [classes/:?]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:78) [classes/:?]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:192) [classes/:?]
```

but I agree that tests for the handler would be useful

Reply:
I've added tests for the `CommandApiRequestHandler` but not for `CommandApiRequestReader`.